### ISSUE TITLE
Debug app: allow multiple instances

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#948]
- [ ] `quick-lint-js --list-debug-app-servers` to list active server host+ports

**Comment by strager** [comment]
Ideas for how to detect instances:

* lsof on the executable file
* ps on processes named "quick-lint-js"
  * Problem: Won't detect VS Code extensions
* port scan
  * Problem: Might hang
  * Problem: Might affect other services
* register PIDs in a PID file somewhere (~/.cache?)
  * Problem: How to garbage collect?
  * Problem: Concurrency

**Comment by strager** [comment]
I should look into how VS Code does it for their `code` command-line utility.

**Comment by strager** [comment]
VS Code uses a UDS or Win32 named pipe: https://github.com/microsoft/vscode/blob/ef1542d820532b7f5ed31c785b31d8f8401c3cb3/src/vs/base/parts/ipc/node/ipc.net.ts#L786

The CLI tries to create a server. If it already exists, it is instead connected to: https://github.com/microsoft/vscode/blob/ef1542d820532b7f5ed31c785b31d8f8401c3cb3/src/vs/code/electron-main/main.ts#L269-L271

**Comment by strager** [comment]
Proposed algorithms for PID registration:
```
onStart {
  p = open("pidfile", O_APPEND|O_WRONLY)
  write(p, "${mypid} ${myport}\n")  // atomic
  close(p)
}

onExit {
  o = open("pidfile", O_RDONLY)
  old_pids = read(o)
  alive_pids = remove old_pids which are not alive
try_again:
  n = open("newpidfile", O_WRONLY)
  write(n, alive_pids)
  close(n)

  rename("newpidfile", "pidfile")  // atomic overwrite
  old_pids2 = read(o)
  if old_pids != old_pids2 {
    alive_pids2 = remove old_pids which are not alive
    if alive_pids != alive_pids2 {
      // Somebody added a PID while we were garbage collecting.
      old_pids = old_pids2
      alive_pids = alive_pids2
      goto try_again
    }
  }
}
```

Problem with the above: If a process starts, its PID might be dropped if an exiting process crashes.
Problem with the above: It's tricky.

**Comment by strager** [comment]
New idea: One file per process (`~/.cache/quick-lint-js.${PID}`).
* Start: Create and truncate file
* Stop: Delete file
* GC (on stop): Look for other PID files and just delete them
* Scan: Look for PID files

**Comment by strager** [comment]
We can detect these things:
* magic in memory (SLOW)
* memory mapping tag?
* thread name?
* open files
  * can store data in file offset
* open UDS
* open tty
* open TCP and UDP ports

**Comment by strager** [comment]
twitch.tv/qyxat made a proposal using a shared file: https://bpa.st/ZWOY4

**Comment by strager** [comment]
Querying threads seems to work well on Linux and macOS. I'll try Windows soon.

**Comment by strager** [comment]
Implemented the --debug-apps CLI option in commit d1d4569e632a9bf8326da35866931a596e330b3a. Supported randomized port in commit c2563a0bc489c8f8d5f4d79c5cdd412b762bf3ff.
